### DISCUSSION TITLE
nextjs render issue

### DISCUSSION THREAD

**Discussion by Kobe** [vercel/next.js#37743] [Help] [answered]
Hi, I am currently facing an odd render issue with integrating a component which returns another component / html element via mapping. If you pass an invalid property the component has an undefined return value.

Does anyone know how to handle this correctly?
Why is nextjs not reliable at this point by ignoring this component during rendering?

Example: https://github.com/Kobe/nextjs-undefined-component/blob/main/pages/index.js
Error: `Server Error
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
This error happened while generating the page. Any console logs will be displayed in the terminal window.`

Thanks in advance!

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I think you might be confused here, the problem is not necessarily with Next.js but it is a problem with understanding what JSX really is. 

> This is very well covered in various courses and online posts, to a far better extend than I could before I go out for my lunch break, so if I am not clear enough, apologies, and feel free to ask any follow ups. Usually when giving a React talk this is the part I practice most, and keep coming back to because it is such an essential building block of working with React that I feel like using the wrong word might cause total confusion.

First let's see a fix. For people who use TypeScript, this wouldn't been natural but, we have to account for what happens when we have a key that doesn't exist in the look up object.

```jsx
const ComponentWithMapping = ({ value }) => {
  if (!value) {
    return null;
  }

  const COMPONENT_WITH_MAPPING = MAPPING[value];

  if (!COMPONENT_WITH_MAPPING) return COMPONENT_WITH_MAPPING; // more on this near the end

  return <COMPONENT_WITH_MAPPING />;
};
```

Or

```
 if (!COMPONENT_WITH_MAPPING) return null;
```

Why is this though? Well, we need to understand what JSX is, and what's the difference between:

```jsx
let element = <div>Hello world</div>
```

and 

```jsx
let Element = () => <div>Hello world</div>
```

`element` is a JSX object, and `Element` is a React Component that returns a JSX object. JSX objects are often called JSX Elements. 

When we do `<Element />`, we are rendering the component, not calling the function, as in `Element()`. In fact, the latter would cause a few issues of its own.

Imagine if you tried to do this:

```jsx
function App() {
  const MyComponent = undefined;
  return <MyComponent />;
}
```

It would fail, same for `null` and a bunch of other values, it is only for certain strings, that resolve to native DOM nodes, be it, `div`, `span`, `p`, etc, and for Functions that return JSX, or Classes which extend React.Component or React.PureComponent. 

You'd only find the strings mapping to DOM nodes limitation in TypeScript. JavaScript and HTML are more forgiving and will let you render any tag you wish.

```jsx
function App() {
  const Tag = 'button';
  return <Tag>Hello</Tag>;
}
```

Under the hood all of the `<JSX />` expressions boil down to (though this has changed slightly in React 17)

```jsx
ui = <Capitalized /> // React.createElement(Capitalized)
ui = <property.access /> // React.createElement(property.access)
ui = <Property.Access /> // React.createElement(Property.Access)
ui = <Property['Access'] /> // SyntaxError
ui = <lowercase /> // React.createElement('lowercase')
ui = <kebab-case /> // React.createElement('kebab-case')
ui = <Upper-Kebab-Case /> // React.createElement('Upper-Kebab-Case')
ui = <Upper_Snake_Case /> // React.createElement(Upper_Snake_Case)
ui = <lower_snake_case /> // React.createElement('lower_snake_case')
```

Now, let's not be confused with the return types. You can return `null` from a React Component, or a JSX.Element, and until React 18, you couldn't return `undefined`, but in React 18 you can. https://blog.saeloun.com/2022/04/14/react-18-allows-components-to-render-undfined.html

Whatever goes between `<>` has to either be a Class that renders JSX, or a Function that returns JSX, or a string representing a DOM element.

This also fails in Client Side Rendering, btw, CRA, or Vite, will run into the same error, because, as I have tried to show, the issue is that when we create a React element, it has to be of a particular type, a string, or a Function/Class that when executed returns JSX, null or `undefined`.

Just for completeness a TypeScript fix would look like:

```tsx
import Head from "next/head";

const MAPPING = {
  A: () => <h1>A</h1>,
  B: () => <h1>B</h1>,
  C: () => <h1>C</h1>,
};

const hasComponent = (value: string): value is keyof typeof MAPPING =>
  value in MAPPING;

const ComponentWithMapping = ({ value }: { value: string }) => {
  if (!value) {
    return null;
  }

  if (hasComponent(value)) {
    const Tag = MAPPING[value];
    return <Tag />;
  }

  return null;
};

export default function Home() {
  return (
    <div>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main>
        <ComponentWithMapping value="D" />
      </main>
    </div>
  );
}
```

TypeScript would've stopped you before you could even render.


### Questions

1. Does anyone know how to handle this correctly?

Check if value exists in the map, and if so, do as you did, otherwise just return `null`, without constructing a Component, just plain `null`

2. Why is nextjs not reliable at this point by ignoring this component during rendering?

The example is trying to use `undefined` as a string, function that returns JSX or class that renders JSX. It is React that's not accepting this situation because of JSX semantics.

**Reply by Kobe** [reply]
@icyJoseph  thank you so much for the extended explanation - appreciate it. Now I have a better understanding what went wrong.

My conclusion: Due to the nature of React, those scenarios must be handled right into the components.
